File: modules/expo-async-fs/ios/ExpoAsyncFsModule.swift
import ExpoModulesCore
import Foundation

public class ExpoAsyncFsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoAsyncFs")

    Events("onDownloadProgress")

    AsyncFunction("listDirectoryAsync") { (uri: String) -> [String] in
      return try FileManager.default.contentsOfDirectory(atPath: Self.resolvePath(uri))
    }

    AsyncFunction("getDirectorySizeAsync") { (uri: String) -> Int in
      return Self.directorySize(at: Self.fileUrl(uri))
    }

    // One off-thread call that returns each entry's name + size + type, so
    // callers avoid a sync .exists/.size stat per child on the JS thread
    // (expo-file-system's .exists/.size are sync-only). Used by
    // reconcileImageCache to walk the cover-art cache without blocking JS.
    AsyncFunction("listDirectoryWithSizesAsync") { (uri: String) -> [[String: Any]] in
      let url = Self.fileUrl(uri)
      let fm = FileManager.default
      let names = try fm.contentsOfDirectory(atPath: url.path)
      return names.map { name in
        let childPath = url.appendingPathComponent(name).path
        var isDir: ObjCBool = false
        fm.fileExists(atPath: childPath, isDirectory: &isDir)
        let size: Int = isDir.boolValue
          ? 0
          : ((try? fm.attributesOfItem(atPath: childPath)[.size] as? Int) ?? 0)
        return [
          "name": name,
          "size": size,
          "isDirectory": isDir.boolValue,
        ]
      }
    }

    // Off-thread existence + size + type stat. Single call so a render-path
    // consumer can confirm a file without a sync .exists/.size on the JS
    // thread (expo-file-system's are sync-only). `size` is 0 for missing
    // entries and directories.
    AsyncFunction("statAsync") { (uri: String) -> [String: Any] in
      let path = Self.resolvePath(uri)
      let fm = FileManager.default
      var isDir: ObjCBool = false
      let exists = fm.fileExists(atPath: path, isDirectory: &isDir)
      let size: Int = (exists && !isDir.boolValue)
        ? ((try? fm.attributesOfItem(atPath: path)[.size] as? Int) ?? 0)
        : 0
      return [
        "exists": exists,
        "size": size,
        "isDirectory": isDir.boolValue,
      ]
    }

    // Off-thread file delete. Returns true if a file existed and was deleted.
    AsyncFunction("deleteFileAsync") { (uri: String) -> Bool in
      let path = Self.resolvePath(uri)
      let fm = FileManager.default
      guard fm.fileExists(atPath: path) else { return false }
      try fm.removeItem(atPath: path)
      return true
    }

    // Off-thread RECURSIVE directory delete (whole cache wipe on logout /
    // clear-cache). expo-file-system's Directory.delete is sync-only and would
    // unlink potentially thousands of files on the JS thread. FileManager's
    // removeItem is recursive for directories.
    AsyncFunction("deleteDirectoryAsync") { (uri: String) -> Bool in
      let path = Self.resolvePath(uri)
      let fm = FileManager.default
      guard fm.fileExists(atPath: path) else { return false }
      try fm.removeItem(atPath: path)
      return true
    }

    AsyncFunction("downloadFileAsyncWithProgress") { (urlString: String, destinationUri: String, downloadId: String) -> [String: Any] in
      guard let url = URL(string: urlString) else {
        throw DownloadError.invalidUrl
      }
      // Remote URL above keeps URL(string:); the destination is a local file
      // path, so resolve it via the space-tolerant resolver (a literal space
      // would make URL(string:) nil and fail every download to such a path).
      let destUrl = Self.fileUrl(destinationUri)

      var request = URLRequest(url: url)
      request.cachePolicy = .reloadIgnoringLocalCacheData

      let config = URLSessionConfiguration.default
      config.requestCachePolicy = .reloadIgnoringLocalCacheData
      config.urlCache = nil

      var lastEventTime: TimeInterval = 0
      let delegate = DownloadProgressDelegate(
        destinationUrl: destUrl,
        onProgress: { [weak self] bytesWritten, totalBytes in
          let now = ProcessInfo.processInfo.systemUptime
          let isComplete = totalBytes > 0 && bytesWritten >= totalBytes
          guard now - lastEventTime >= 0.1 || isComplete else { return }
          lastEventTime = now
          self?.sendEvent("onDownloadProgress", [
            "downloadId": downloadId,
            "bytesWritten": bytesWritten,
            "totalBytes": totalBytes,
          ])
        }
      )

      let session = URLSession(
        configuration: config,
        delegate: delegate,
        delegateQueue: nil
      )

      defer { session.finishTasksAndInvalidate() }

      try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        delegate.continuation = continuation
        session.downloadTask(with: request).resume()
      }

      let fileSize = (try? FileManager.default.attributesOfItem(atPath: destUrl.path)[.size] as? Int64) ?? 0

      return [
        "uri": destUrl.absoluteString,
        "bytes": fileSize,
      ]
    }
  }

  /// Resolve a file URI (or bare path) to a filesystem path. Mirrors the
  /// expo-image-resize resolver: URL(string:) percent-decodes a well-formed
  /// file:// URI, but returns nil when the URI carries a literal (unencoded)
  /// space — in which case we strip the scheme so paths with spaces still
  /// resolve instead of silently failing the whole operation. A bare path
  /// (no scheme) is returned unchanged.
  private static func resolvePath(_ uri: String) -> String {
    if uri.hasPrefix("file://") {
      return URL(string: uri)?.path ?? String(uri.dropFirst("file://".count))
    }
    return uri
  }

  /// Convenience for the call sites that need a file URL rather than a path.
  private static func fileUrl(_ uri: String) -> URL {
    return URL(fileURLWithPath: resolvePath(uri))
  }

  private static func directorySize(at url: URL) -> Int {
    let fm = FileManager.default
    guard let enumerator = fm.enumerator(
      at: url,
      includingPropertiesForKeys: [.fileSizeKey],
      options: [.skipsHiddenFiles]
    ) else { return 0 }

    var total = 0
    for case let fileURL as URL in enumerator {
      total += (try? fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
    }
    return total
  }
}

private enum DownloadError: Error, LocalizedError {
  case invalidUrl
  case invalidDestination
  case httpError(Int)

  var errorDescription: String? {
    switch self {
    case .invalidUrl: return "Invalid download URL"
    case .invalidDestination: return "Invalid destination path"
    case .httpError(let code): return "Download failed with HTTP status \(code)"
    }
  }
}

/// Uses URLSession.downloadTask with a delegate — the same download
/// mechanism as expo-file-system's File.downloadFileAsync (which uses
/// the completion-handler variant). The delegate approach gives us the
/// additional didWriteData callback for progress events.
///
/// The temp file is moved to the destination inside didFinishDownloadingTo,
/// matching how expo-file-system moves it inside the completion handler.
/// iOS deletes the temp file once the callback returns, so the move
/// must happen before then.
private class DownloadProgressDelegate: NSObject, URLSessionDownloadDelegate {
  let destinationUrl: URL
  let onProgress: (Int64, Int64) -> Void
  var continuation: CheckedContinuation<Void, Error>?

  init(destinationUrl: URL, onProgress: @escaping (Int64, Int64) -> Void) {
    self.destinationUrl = destinationUrl
    self.onProgress = onProgress
  }

  func urlSession(
    _ session: URLSession,
    downloadTask: URLSessionDownloadTask,
    didWriteData bytesWritten: Int64,
    totalBytesWritten: Int64,
    totalBytesExpectedToWrite: Int64
  ) {
    onProgress(totalBytesWritten, totalBytesExpectedToWrite)
  }

  func urlSession(
    _ session: URLSession,
    downloadTask: URLSessionDownloadTask,
    didFinishDownloadingTo location: URL
  ) {
    let statusCode = (downloadTask.response as? HTTPURLResponse)?.statusCode ?? 200
    guard statusCode >= 200 && statusCode < 300 else {
      continuation?.resume(throwing: DownloadError.httpError(statusCode))
      continuation = nil
      return
    }

    do {
      if FileManager.default.fileExists(atPath: destinationUrl.path) {
        try FileManager.default.removeItem(at: destinationUrl)
      }
      try FileManager.default.moveItem(at: location, to: destinationUrl)
      continuation?.resume(returning: ())
    } catch {
      continuation?.resume(throwing: error)
    }
    continuation = nil
  }

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    guard let error = error else { return }
    continuation?.resume(throwing: error)
    continuation = nil
  }
}

File: modules/expo-async-fs/android/src/main/java/expo/modules/asyncfs/ExpoAsyncFsModule.kt
package expo.modules.asyncfs

import android.net.Uri
import android.os.Bundle
import androidx.core.os.bundleOf
import com.facebook.react.modules.network.OkHttpClientProvider
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import okhttp3.MediaType
import okhttp3.Request
import okhttp3.ResponseBody
import okio.Buffer
import okio.BufferedSource
import okio.ForwardingSource
import okio.Source
import okio.buffer
import java.io.File
import java.io.FileOutputStream

class ExpoAsyncFsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoAsyncFs")

    Events("onDownloadProgress")

    AsyncFunction("listDirectoryAsync") { uri: String ->
      val path = Uri.parse(uri).path ?: return@AsyncFunction emptyList<String>()
      File(path).list()?.toList() ?: emptyList()
    }

    AsyncFunction("getDirectorySizeAsync") { uri: String ->
      val path = Uri.parse(uri).path ?: return@AsyncFunction 0L
      directorySize(File(path))
    }

    // One off-thread call that returns each entry's name + size + type, so
    // callers don't have to do a sync File.exists/File.size stat per child on
    // the JS thread (expo-file-system's .exists/.size are sync-only). Used by
    // reconcileImageCache to walk the cover-art cache without blocking JS.
    AsyncFunction("listDirectoryWithSizesAsync") { uri: String ->
      val path = Uri.parse(uri).path ?: return@AsyncFunction emptyList<Bundle>()
      val entries = File(path).listFiles() ?: return@AsyncFunction emptyList<Bundle>()
      entries.map { f ->
        bundleOf(
          "name" to f.name,
          "size" to (if (f.isFile) f.length().toDouble() else 0.0),
          "isDirectory" to f.isDirectory,
        )
      }
    }

    // Off-thread existence + size + type stat. Single call so a render-path
    // consumer can confirm a file without a sync File.exists/File.length on the
    // JS thread (expo-file-system's are sync-only). `size` is 0 for missing
    // entries and directories.
    AsyncFunction("statAsync") { uri: String ->
      val path = Uri.parse(uri).path
        ?: return@AsyncFunction bundleOf("exists" to false, "size" to 0.0, "isDirectory" to false)
      val f = File(path)
      val exists = f.exists()
      bundleOf(
        "exists" to exists,
        "size" to (if (exists && f.isFile) f.length().toDouble() else 0.0),
        "isDirectory" to (exists && f.isDirectory),
      )
    }

    // Off-thread file delete. Returns true if a file existed and was deleted.
    AsyncFunction("deleteFileAsync") { uri: String ->
      val path = Uri.parse(uri).path ?: return@AsyncFunction false
      val f = File(path)
      if (f.exists()) f.delete() else false
    }

    // Off-thread RECURSIVE directory delete (the whole cache wipe on logout /
    // clear-cache). expo-file-system's Directory.delete is sync-only and would
    // unlink potentially thousands of files on the JS thread. Dispatched to
    // Dispatchers.IO so a large wipe runs in parallel off the module queue.
    AsyncFunction("deleteDirectoryAsync") { uri: String, promise: Promise ->
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val path = Uri.parse(uri).path
          if (path == null) {
            promise.resolve(false)
            return@launch
          }
          val dir = File(path)
          promise.resolve(if (dir.exists()) dir.deleteRecursively() else false)
        } catch (e: Exception) {
          promise.reject("ERR_DELETE_DIR", e.message ?: "Recursive delete failed", e)
        }
      }
    }

    // Mirrors expo-file-system's downloadFileAsync but adds a network
    // interceptor for progress events. Uses OkHttpClientProvider (rather
    // than a bare OkHttpClient) so the RN network stack configuration
    // (including custom SSL trust) is inherited.
    //
    // Takes a Promise parameter and dispatches to Dispatchers.IO so that
    // concurrent calls run on separate threads. Expo's default module
    // queue is a single HandlerThread; without this, blocking execute()
    // calls would serialize all downloads.
    AsyncFunction("downloadFileAsyncWithProgress") { url: String, destinationUri: String, downloadId: String, promise: Promise ->
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val destPath = Uri.parse(destinationUri).path
            ?: throw Exception("Invalid destination URI")
          val destFile = File(destPath)
          destFile.parentFile?.mkdirs()

          var lastEventTime = 0L

          val client = OkHttpClientProvider.createClient().newBuilder()
            .addNetworkInterceptor { chain ->
              val response = chain.proceed(chain.request())
              val body = response.body ?: return@addNetworkInterceptor response
              val contentLength = body.contentLength()
              val source = object : ForwardingSource(body.source()) {
                var totalBytesRead = 0L

                override fun read(sink: Buffer, byteCount: Long): Long {
                  val bytesRead = super.read(sink, byteCount)
                  if (bytesRead != -1L) totalBytesRead += bytesRead
                  val now = System.currentTimeMillis()
                  val isComplete = contentLength > 0 && totalBytesRead >= contentLength
                  if (now - lastEventTime >= 100 || isComplete) {
                    lastEventTime = now
                    sendEvent("onDownloadProgress", bundleOf(
                      "downloadId" to downloadId,
                      "bytesWritten" to totalBytesRead.toDouble(),
                      "totalBytes" to contentLength.toDouble(),
                    ))
                  }
                  return bytesRead
                }
              }
              response.newBuilder()
                .body(ProgressResponseBody(body.contentType(), contentLength, source.buffer()))
                .build()
            }
            .build()

          val request = Request.Builder().url(url).build()
          val response = client.newCall(request).execute()

          if (!response.isSuccessful) {
            response.close()
            throw Exception("Download failed with HTTP status ${response.code}")
          }

          val body = response.body ?: throw Exception("Empty response body")
          // Completion is judged by HTTP status + a clean read to EOF, NOT by
          // matching bytes-written against Content-Length. Subsonic estimates
          // Content-Length for on-the-fly transcodes, so the real byte count
          // legitimately differs from the header — a length check would reject
          // good downloads. A genuinely incomplete transfer surfaces as a
          // non-success status (handled above) or a stream read error thrown
          // out of copyTo below; reaching the end cleanly on a 2xx response is
          // authoritative completion.
          body.byteStream().use { input ->
            FileOutputStream(destFile).use { output ->
              input.copyTo(output)
            }
          }

          val fileSize = destFile.length()
          promise.resolve(bundleOf(
            "uri" to Uri.fromFile(destFile).toString(),
            "bytes" to fileSize.toDouble(),
          ))
        } catch (e: Exception) {
          promise.reject("ERR_DOWNLOAD", e.message ?: "Download failed", e)
        }
      }
    }
  }

  private fun directorySize(dir: File): Long {
    if (!dir.exists()) return 0
    return dir.walkTopDown().filter { it.isFile }.sumOf { it.length() }
  }
}

private class ProgressResponseBody(
  private val contentType: MediaType?,
  private val contentLength: Long,
  private val bufferedSource: BufferedSource,
) : ResponseBody() {
  override fun contentType(): MediaType? = contentType
  override fun contentLength(): Long = contentLength
  override fun source(): BufferedSource = bufferedSource
}
